File: examples/python_dashboardv2/raspberry/disk.py
import grafana_foundation_sdk.models.dashboardv2beta1 as dashboard
from grafana_foundation_sdk.builders.common import TableFooterOptions
from grafana_foundation_sdk.builders.dashboardv2beta1 import (
    Panel,
    QueryGroup,
    Target,
    Transformation,
)
from grafana_foundation_sdk.builders.table import Visualization as TableVisualization
from grafana_foundation_sdk.models.common import (
    FieldTextAlignment,
    TableAutoCellOptions,
    TableCellHeight,
)

from .common import default_timeseries, basic_prometheus_query, table_prometheus_query


def disk_io_timeseries() -> Panel:
    return (
        Panel()
        .title("Disk I/O")
        .visualization(
            default_timeseries()
            .fill_opacity(0)
            .unit("Bps")
            .override_by_regexp(
                "/ io time/",
                [
                    dashboard.DynamicConfigValue(id_val="unit", value="percentunit"),
                ],
            )
        )
        .data(
            QueryGroup()
            .target(
                Target().query(
                    basic_prometheus_query(
                        'rate(node_disk_read_bytes_total{job="integrations/raspberrypi-node", instance="$instance", device!=""}[$__rate_interval])',
                        "{{device}} read",
                    )
                ).ref_id("A")
            )
            .target(
                Target().query(
                    basic_prometheus_query(
                        'rate(node_disk_written_bytes_total{job="integrations/raspberrypi-node", instance="$instance", device!=""}[$__rate_interval])',
                        "{{device}} written",
                    )
                ).ref_id("B")
            )
            .target(
                Target().query(
                    basic_prometheus_query(
                        'rate(node_disk_io_time_seconds_total{job="integrations/raspberrypi-node", instance="$instance", device!=""}[$__rate_interval])',
                        "{{device}} IO time",
                    )
                ).ref_id("C")
            )
        )
    )


def disk_space_usage_table() -> Panel:
    return (
        Panel()
        .title("Disk Space Usage")
        .visualization(
            TableVisualization()
            .align(FieldTextAlignment.AUTO)
            .cell_options(TableAutoCellOptions())
            .unit("decbytes")
            .cell_height(TableCellHeight.SM)
            .footer(TableFooterOptions().count_rows(False).reducer(["sum"]))
            # Overrides configuration
            .override_by_name(
                "Mounted on",
                [
                    dashboard.DynamicConfigValue(id_val="custom.width", value=260),
                ],
            )
            .override_by_name(
                "Size",
                [
                    dashboard.DynamicConfigValue(id_val="custom.width", value=93),
                ],
            )
            .override_by_name(
                "Used",
                [
                    dashboard.DynamicConfigValue(id_val="custom.width", value=72),
                ],
            )
            .override_by_name(
                "Available",
                [
                    dashboard.DynamicConfigValue(id_val="custom.width", value=88),
                ],
            )
            .override_by_name(
                "Used, %",
                [
                    dashboard.DynamicConfigValue(id_val="unit", value="percentunit"),
                    dashboard.DynamicConfigValue(
                        id_val="custom.cellOptions",
                        value={"mode": "gradient", "type": "gauge"},
                    ),
                    dashboard.DynamicConfigValue(id_val="max", value="1"),
                    dashboard.DynamicConfigValue(id_val="min", value="0"),
                ],
            )
        )
        .data(
            QueryGroup()
            .target(
                Target().query(
                    table_prometheus_query(
                        'max by (mountpoint) (node_filesystem_size_bytes{job="integrations/raspberrypi-node", instance="$instance", fstype!=""})'
                    )
                ).ref_id("A")
            )
            .target(
                Target().query(
                    table_prometheus_query(
                        'max by (mountpoint) (node_filesystem_avail_bytes{job="integrations/raspberrypi-node", instance="$instance", fstype!=""})',
                    )
                ).ref_id("B")
            )
            # Transformations
            .transformation(
                Transformation()
                .id("groupBy")
                .kind("groupBy")
                .options(
                    {
                        "fields": {
                            "Value #A": {
                                "aggregations": ["lastNotNull"],
                                "operation": "aggregate",
                            },
                            "Value #B": {
                                "aggregations": ["lastNotNull"],
                                "operation": "aggregate",
                            },
                            "mountpoint": {"aggregations": [], "operation": "groupby"},
                        }
                    }
                )
            )
            .transformation(Transformation().id("merge").kind("merge").options({}))
            .transformation(
                Transformation()
                .id("calculateField")
                .kind("calculateField")
                .options(
                    {
                        "alias": "Used",
                        "binary": {
                            "left": "Value #A (lastNotNull)",
                            "operator": "-",
                            "reducer": "sum",
                            "right": "Value #B (lastNotNull)",
                        },
                        "mode": "binary",
                        "reduce": {"reducer": "sum"},
                    }
                )
            )
            .transformation(
                Transformation()
                .id("calculateField")
                .kind("calculateField")
                .options(
                    {
                        "alias": "Used, %",
                        "binary": {
                            "left": "Used",
                            "operator": "/",
                            "reducer": "sum",
                            "right": "Value #A (lastNotNull)",
                        },
                        "mode": "binary",
                        "reduce": {"reducer": "sum"},
                    }
                )
            )
            .transformation(
                Transformation()
                .id("organize")
                .kind("organize")
                .options(
                    {
                        "excludeByName": {},
                        "indexByName": {},
                        "renameByName": {
                            "Value #A (lastNotNull)": "Size",
                            "Value #B (lastNotNull)": "Available",
                            "mountpoint": "Mounted on",
                        },
                    }
                )
            )
            .transformation(
                Transformation()
                .id("sortBy")
                .kind("sortBy")
                .options({"fields": {}, "sort": [{"field": "Mounted on"}]})
            )
        )
    )

File: examples/typescript_dashboardv2/disk.ts
import {VisualizationBuilder as TableVizBuilder} from "../../generated/typescript/src/table";
import {basicPrometheusQuery, defaultTimeseries, tablePrometheusQuery} from "./common";
import {
    FieldTextAlignment,
    TableCellDisplayMode,
    TableCellHeight,
    TableFooterOptionsBuilder
} from "../../generated/typescript/src/common";
import {
    PanelBuilder,
    QueryGroupBuilder,
    TargetBuilder,
    ThresholdsConfigBuilder,
    ThresholdsMode,
    TransformationBuilder,
} from "../../generated/typescript/src/dashboardv2beta1";

export const diskIOTimeseries = (): PanelBuilder => {
    return new PanelBuilder()
        .title("Disk I/O")
        .visualization(defaultTimeseries()
            .fillOpacity(0)
            .unit("Bps")
            .thresholds(
                new ThresholdsConfigBuilder()
                    .mode(ThresholdsMode.Absolute)
                    .steps([
                        {value: null, color: "green"},
                        {value: 80.0, color: "red"},
                    ])
            )
            .overrideByRegexp("/ io time/", [
                {id: "unit", value: "percentunit"},
            ])
        )
        .data(new QueryGroupBuilder()
            .targets([
                new TargetBuilder().query(basicPrometheusQuery(`rate(node_disk_read_bytes_total{job="integrations/raspberrypi-node", instance="$instance", device!=""}[$__rate_interval])`, "{{device}} read").refId("A")),
                new TargetBuilder().query(basicPrometheusQuery(`rate(node_disk_written_bytes_total{job="integrations/raspberrypi-node", instance="$instance", device!=""}[$__rate_interval])`, "{{device}} written").refId("B")),
                new TargetBuilder().query(basicPrometheusQuery(`rate(node_disk_io_time_seconds_total{job="integrations/raspberrypi-node", instance="$instance", device!=""}[$__rate_interval])`, "{{device}} IO time").refId("C")),
            ])
        );
};

export const diskSpaceUsageTable = (): PanelBuilder => {
    return new PanelBuilder()
        .title("Disk Space Usage")
        .visualization(new TableVizBuilder()
            .align(FieldTextAlignment.Auto)
            .cellOptions({type: TableCellDisplayMode.Auto})
            .cellHeight(TableCellHeight.Sm)
            .footer(new TableFooterOptionsBuilder().countRows(false).reducer(["sum"]))
            .unit("decbytes")

            // Overrides configuration
            .overrideByName("Mounted on",[
                {id: "custom.width", value: 260},
            ])
            .overrideByName("Size", [
                {id: "custom.width", value: 93},
            ])
            .overrideByName("Used", [
                {id: "custom.width", value: 72},
            ])
            .overrideByName("Available", [
                {id: "custom.width", value: 88},
            ])
            .overrideByName("Used, %", [
                {id: "unit", value: "percentunit"},
                {
                    id: "custom.cellOptions",
                    value: {mode: "gradient", type: "gauge"}
                },
                {id: "max", value: 1},
                {id: "min", value: 0}
            ])
        )
        .data(new QueryGroupBuilder()
            .targets([
                new TargetBuilder().query(tablePrometheusQuery(`max by (mountpoint) (node_filesystem_size_bytes{job="integrations/raspberrypi-node", instance="$instance", fstype!=""})`).refId("A")),
                new TargetBuilder().query(tablePrometheusQuery(`max by (mountpoint) (node_filesystem_avail_bytes{job="integrations/raspberrypi-node", instance="$instance", fstype!=""})`).refId("B")),
            ])
            // TODO: transformations are very clunky
            .transformation(new TransformationBuilder()
                .kind("groupBy")
                .id("groupBy")
                .options({
                    fields: {
                        "Value #A": {
                            aggregations: ["lastNotNull"],
                            operation: "aggregate"
                        },
                        "Value #B": {
                            aggregations: ["lastNotNull"],
                            operation: "aggregate"
                        },
                        mountpoint: {
                            aggregations: [],
                            operation: "groupby"
                        }
                    }
                })
            )
            .transformation(new TransformationBuilder()
                .kind("merge")
                .id("merge")
                .options({})
            )
            .transformation(new TransformationBuilder()
                .kind("calculateField") // is `kind` the type of transformation?
                .id("calculateField") // should this be different from `kind`?
                .options({
                    alias: "Used",
                    binary: {
                        left: "Value #A (lastNotNull)",
                        operator: "-",
                        reducer: "sum",
                        right: "Value #B (lastNotNull)"
                    },
                    mode: "binary",
                    reduce: {reducer: "sum"}
                })
            )
            .transformation(new TransformationBuilder()
                .kind("calculateField")
                .id("calculateField")
                .options({
                    alias: "Used, %",
                    binary: {
                        left: "Used",
                        operator: "/",
                        reducer: "sum",
                        right: "Value #A (lastNotNull)"
                    },
                    mode: "binary",
                    reduce: {"reducer": "sum"}
                })
            )
            .transformation(new TransformationBuilder()
                .kind("organize")
                .id("organize")
                .options({
                    excludeByName: {},
                    indexByName: {},
                    renameByName: {
                        "Value #A (lastNotNull)": "Size",
                        "Value #B (lastNotNull)": "Available",
                        mountpoint: "Mounted on"
                    }
                })
            )
            .transformation(new TransformationBuilder()
                .kind("sortBy")
                .id("sortBy")
                .options({
                    fields: {},
                    sort: [
                        {field: "Mounted on"}
                    ]
                })
            )
        );
};
